crates/net/discv5/src/config.rs

```diff
@@ -286,12 +312,104 @@ impl Config {
-        match self.discv5_config.listen_config {
-            ListenConfig::Ipv4 { ip, .. } => (ip, port).into(),
-            ListenConfig::Ipv6 { ip, .. } => (ip, port).into(),
-            ListenConfig::DualStack { ipv4, .. } => (ipv4, port).into(),
+    pub fn rlpx_socket(&self) -> &SocketAddr {
+        &self.tcp_socket
+    }
+}
+
+/// Returns the IPv4 discovery socket if one is configured.
+pub fn ipv4(listen_config: &ListenConfig) -> Option<SocketAddrV4> {
+    match listen_config {
+        ListenConfig::Ipv4 { ip, port } |
+        ListenConfig::DualStack { ipv4: ip, ipv4_port: port, .. } => {
+            Some(SocketAddrV4::new(*ip, *port))
+        }
+        ListenConfig::Ipv6 { .. } => None,
+    }
+}
+
+/// Returns the IPv6 discovery socket if one is configured.
+pub fn ipv6(listen_config: &ListenConfig) -> Option<SocketAddrV6> {
+    match listen_config {
+        ListenConfig::Ipv4 { .. } => None,
+        ListenConfig::Ipv6 { ip, port } |
+        ListenConfig::DualStack { ipv6: ip, ipv6_port: port, .. } => {
+            Some(SocketAddrV6::new(*ip, *port, 0, 0))
+        }
+    }
+}
+
+/// Returns the amended [`discv5::ListenConfig`] based on the RLPx IP address. The ENR is limited
+/// to one IP address per IP version (atm, may become spec'd how to advertise different addresses).
+/// The RLPx address overwrites the discv5 address w.r.t. IP version.
+pub fn amend_listen_config_wrt_rlpx(
+    rlpx_addr: IpAddr,
+    listen_config: &ListenConfig,
+) -> ListenConfig {
+    let discv5_socket_ipv4 = ipv4(listen_config);
+    let discv5_socket_ipv6 = ipv6(listen_config);
+
+    let discv5_port_ipv4 =
+        discv5_socket_ipv4.map(|socket| socket.port()).unwrap_or(DEFAULT_DISCOVERY_V5_PORT);
+    let discv5_addr_ipv4 = discv5_socket_ipv4.map(|socket| *socket.ip());
+    let discv5_port_ipv6 =
+        discv5_socket_ipv6.map(|socket| socket.port()).unwrap_or(DEFAULT_DISCOVERY_V5_PORT);
+    let discv5_addr_ipv6 = discv5_socket_ipv6.map(|socket| *socket.ip());
+
+    let (discv5_socket_ipv4, discv5_socket_ipv6) = discv5_sockets_wrt_rlpx_addr(
+        rlpx_addr,
+        discv5_addr_ipv4,
+        discv5_port_ipv4,
+        discv5_addr_ipv6,
+        discv5_port_ipv6,
+    );
+
+    ListenConfig::from_two_sockets(discv5_socket_ipv4, discv5_socket_ipv6)
+}
+
+/// Returns the sockets that can be used for discv5 with respect to RLPx address. ENR specs only
+/// acknowledge one address per IP version.
+pub fn discv5_sockets_wrt_rlpx_addr(
+    rlpx_addr: IpAddr,
+    discv5_addr_ipv4: Option<Ipv4Addr>,
+    discv5_port_ipv4: u16,
+    discv5_addr_ipv6: Option<Ipv6Addr>,
+    discv5_port_ipv6: u16,
+) -> (Option<SocketAddrV4>, Option<SocketAddrV6>) {
+    match rlpx_addr {
+        IpAddr::V4(rlpx_addr) => {
+            let discv5_socket_ipv6 =
+                discv5_addr_ipv6.map(|ip| SocketAddrV6::new(ip, discv5_port_ipv6, 0, 0));
+
+            if let Some(discv5_addr) = discv5_addr_ipv4 {
+                warn!(target: "reth::cli",
```

Comment:
hmm, warn! should get displayed in the reth cli I believe, so I think we can change the target here and still have it logged to stderr in reth